src/resolver.rs

```diff
@@ -700,27 +700,50 @@ impl<'i> TypeAnnotator<'i> {
                 self.update_expected_types(expected_type, initializer);
 
                 // handle annotation for array of struct
-                if let DataTypeInformation::Array { inner_type_name, .. } =
-                    expected_type.get_type_information()
-                {
-                    let struct_type = self.index.get_effective_type_or_void_by_name(inner_type_name);
-                    if struct_type.get_type_information().is_struct() {
-                        if let AstStatement::ExpressionList { expressions, .. } = initializer {
-                            let ctx = ctx.with_qualifier(struct_type.get_name().to_string());
-                            for e in expressions {
-                                // annotate with the arrays inner_type
-                                self.annotation_map.annotate_type_hint(
-                                    e,
-                                    StatementAnnotation::Value {
-                                        resulting_type: struct_type.get_name().to_string(),
-                                    },
-                                );
-                                self.visit_statement(&ctx, e);
-                            }
+                self.annotate_array_of_struct(expected_type, initializer, &ctx);
+            }
+        }
+    }
+
+    fn annotate_array_of_struct(
+        &mut self,
+        expected_type: &typesystem::DataType,
+        initializer: &AstStatement,
+        ctx: &VisitorContext,
+    ) {
+        match expected_type.get_type_information() {
+            DataTypeInformation::Array { inner_type_name, .. } => {
+                let inner_type = self.index.get_effective_type_or_void_by_name(inner_type_name);
+                let ctx = ctx.with_qualifier(inner_type.get_name().to_string());
+                if inner_type.get_type_information().is_struct() {
+                    if let AstStatement::ExpressionList { expressions, .. } = initializer {
+                        for e in expressions {
```

Comment:
I know it is old code, but can we change this to also use the flatten_expression_list approach you used below

Reply:
This seems to break both array-of-struct as well as array-of-struct-within-a-struct initializers. Right now, for an array of struct we end up with nested expression lists, annotating the struct as a type-hint on the AstStatement::ExpressionList. If we flatten this structure beforehand, we end up annotating the type hint on the AstStatement::Assignment and we are no longer able to generate the initializer during codegen (no type hint on the expression list in `expression_generator::generate_literal`)